src/ArgumentEscaper.cs

```diff
@@ -0,0 +1,207 @@
+    /// </summary>
+    /// <remarks>
+    /// See here for more info: <seealso href="https://docs.microsoft.com/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way"/>.
+    /// </remarks>
+    /// <param name="args">List of arguments to escape.</param>
+    /// <returns>A <c>cmd.exe</c> safe escaped string of the supplied arguments.</returns>
+    public static string EscapeAndConcatenateArgArrayForCmdProcessStart(IReadOnlyList<string>? args)
+    {
+        if (args is null)
+        {
+            return "";
+        }
+
+        return EscapeArgArrayForCmd(args);
+    }
+
+    /// <summary>
+    /// Undo the processing which took place to create string[] args in Main, so that the next process will receive the same string[] args.
+    /// </summary>
+    /// <remarks>
+    /// See here for more info: <seealso href="https://docs.microsoft.com/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way"/>.
+    /// </remarks>
+    /// <param name="arguments">List of arguments to escape.</param>
+    /// <returns>An escaped string of the supplied arguments.</returns>
+    private static string EscapeArgArray(IReadOnlyList<string> arguments)
+    {
+        var sb = new ValueStringBuilder(stackalloc char[256]);
+
+        for (var i = 0; i < arguments.Count; i++)
+        {
+            if (i > 0)
+            {
+                sb.Append(Space);
+            }
+
+            EscapeSingleArg(ref sb, arguments[i]);
+        }
+
+        return sb.ToString();
+    }
+
+    /// <summary>
+    /// This prefixes every character with the '^' character to force cmd to interpret the argument string literally.
+    /// </summary>
+    /// <remarks>
+    /// See here for more info: <seealso href="https://docs.microsoft.com/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way"/>.
+    /// </remarks>
+    /// <param name="arguments">List of arguments to escape.</param>
+    /// <returns>A <c>cmd.exe</c> safe escaped string of the supplied arguments.</returns>
+    private static string EscapeArgArrayForCmd(IReadOnlyList<string> arguments)
+    {
+        var sb = new ValueStringBuilder(stackalloc char[256]);
+
+        for (var i = 0; i < arguments.Count; i++)
+        {
+            if (i > 0)
+            {
+                sb.Append(Space);
+            }
+
+            EscapeArgForCmd(ref sb, arguments[i]);
+        }
+
+        return sb.ToString();
+    }
+
+    private static void EscapeSingleArg(ref ValueStringBuilder sb, string arg)
+    {
+        var length = arg.Length;
+        var needsQuotes = length == 0 || ArgumentContainsWhitespace(arg);
+        var isQuoted = needsQuotes || IsSurroundedWithQuotes(arg);
+
+        if (needsQuotes)
+        {
+            sb.Append(Quote);
+        }
+
+        for (var i = 0; i < length; ++i)
```

Comment:
If you are more consistent with usages of length and args.Length I wonder if that may help with any bounds checks optimizations.

Reply:
That came from the original source, but I removed `length` in favor of `arg.Length` for consistency.